### PR TITLE
TandemFoil: ablate weight decay & EMA for physics+AdamW recipe

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Two unexplored regularization axes for TandemFoil:
1. **Weight decay**: Default is 1e-4. With physics features providing structured inductive bias, the model may need less regularization (lower WD) or more (higher WD to prevent overfitting the physics features).
2. **EMA**: The exponential moving average model (use_ema=True, ema_decay=0.999) is always on by default. For a dataset with ~1300 training cases and 2-5 epochs, EMA may or may not help — the model never sees data many times.

Both interact with the physics + AdamW recipe. Testing them systematically will reveal whether the defaults are optimal.

## Current Baseline

- **val_primary/surface_pressure_mae: 262.82** (PR #2414 — physics + Lion, slices=96, 2 epochs)
- **test_primary/surface_pressure_mae: 257.51**

## Instructions

Run 4 parallel trials with `CUDA_VISIBLE_DEVICES`. Use `--wandb_group radford-tandem-physics-adamw` for all runs. All use physics features + AdamW lr=5e-4, model_slices=64, cosine_t_max=30.

Physics flags for all trials: `--enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition`

### Trial 0 — GPU 0: Default (weight_decay=1e-4, use_ema=True — control)
```bash
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine_t_max 30 --model-slices 64 --weight-decay 1e-4 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb_group radford-tandem-physics-adamw --wandb_name rei/tandem-wd1e4-ema-adamw --agent rei
```

### Trial 1 — GPU 1: Lower weight_decay=1e-5
```bash
CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine_t_max 30 --model-slices 64 --weight-decay 1e-5 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb_group radford-tandem-physics-adamw --wandb_name rei/tandem-wd1e5-ema-adamw --agent rei
```

### Trial 2 — GPU 2: Higher weight_decay=1e-3
```bash
CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine_t_max 30 --model-slices 64 --weight-decay 1e-3 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb_group radford-tandem-physics-adamw --wandb_name rei/tandem-wd1e3-ema-adamw --agent rei
```

### Trial 3 — GPU 3: No EMA (use_ema=False, default weight_decay=1e-4)
```bash
CUDA_VISIBLE_DEVICES=3 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine_t_max 30 --model-slices 64 --weight-decay 1e-4 --no-use-ema --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb_group radford-tandem-physics-adamw --wandb_name rei/tandem-wd1e4-noema-adamw --agent rei
```

Run all 4 in parallel (background each with `&`). Wait for all to finish.

**Key measurements:**
- val_primary/surface_pressure_mae (best across all epochs) for each trial
- Compare Trials 0, 1, 2 to understand weight decay sensitivity
- Compare Trials 0 vs 3 to assess EMA value for TandemFoil training budget

## Results

_To be filled by student after runs complete._

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT rei:

## Results

### Summary Table

| Trial | Weight Decay | EMA | Epochs | val_primary | test_primary |
|-------|-------------|-----|--------|-------------|-------------|
| 0 (control) | 1e-4 | Yes | 2 | 249.97 | 245.31 |
| 1 | 1e-5 | Yes | 1 | 351.24 | 339.91 |
| 2 | 1e-3 | Yes | 2 | 256.23 | 249.69 |
| **3** | **1e-4** | **No** | **1** | **200.89** | **190.17** |

**Baseline:** val=262.82 / test=257.51 (PR #2414, Lion optimizer, slices=96, 2 epochs)

### Per-Split Test Metrics (surface_pressure_mae)

| Split | Trial 0 (wd=1e-4, EMA) | Trial 1 (wd=1e-5, EMA) | Trial 2 (wd=1e-3, EMA) | Trial 3 (wd=1e-4, no EMA) | Baseline |
|-------|------------------------|------------------------|------------------------|--------------------------|----------|
| single_in_dist | 268.21 | 364.82 | 270.91 | **194.80** | 267.26 |
| geom_camber_rc | 239.89 | 417.23 | 271.47 | **186.24** | 280.59 |
| geom_camber_cruise | 227.72 | 253.05 | 212.38 | **180.46** | 225.63 |
| re_rand | 245.40 | 324.53 | 244.02 | **199.19** | 256.55 |
| eq4 (primary) | 245.31 | 339.91 | 249.69 | **190.17** | 257.51 |

### Key Findings

**1. EMA is strongly counterproductive for this short training regime.**
Trial 3 (no EMA, 1 epoch) dramatically outperforms all EMA variants — even those with 2× the training time. With EMA decay=0.999, after 750 steps the EMA model has only absorbed ~53% of gradient updates (1 - 0.999^750 ≈ 0.528). The evaluation model still carries ~47% of the random initialization, which severely degrades predictions. This is the single most impactful finding.

**2. Weight decay: default 1e-4 is best, but sensitivity is moderate.**
Among EMA trials at epoch 2: wd=1e-4 (249.97) slightly beats wd=1e-3 (256.23). At epoch 1: wd=1e-5 (351.24) ≈ wd=1e-4 (356.19) > wd=1e-3 (397.04). Higher weight decay hurts early training, converges similarly later. The comparison is confounded by epoch count (Trial 1 got only 1 epoch).

**3. Trial 3 (no EMA) beats the baseline by 24% on val, 26% on test.**
val: 200.89 vs 262.82 (−61.93), test: 190.17 vs 257.51 (−67.34). Improvements are consistent across all splits (20-34% improvement). Note: this comparison is against a Lion baseline with slices=96; Trial 3 uses AdamW with slices=64, so the improvement reflects both the optimizer change and EMA removal.

**4. All trials were still improving — more epochs would help.**
Loss dropped from ~0.87 (epoch 1) to ~0.46 (epoch 2) for trials that got 2 epochs. The models are far from converged.

### Epoch count note
Trials 0 and 2 completed 2 epochs, while Trials 1 and 3 completed only 1 epoch (30-minute timeout). The timeout checks between epochs, so trials where epoch 1 finished faster could start epoch 2 before the cutoff.

### Commands Used

```bash
# Trial 0 (control): wd=1e-4, EMA on
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine-t-max 30 --model-slices 64 --weight-decay 1e-4 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb-group radford-tandem-physics-adamw --wandb-name rei/tandem-wd1e4-ema-adamw --agent rei

# Trial 1: wd=1e-5, EMA on
CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine-t-max 30 --model-slices 64 --weight-decay 1e-5 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb-group radford-tandem-physics-adamw --wandb-name rei/tandem-wd1e5-ema-adamw --agent rei

# Trial 2: wd=1e-3, EMA on
CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine-t-max 30 --model-slices 64 --weight-decay 1e-3 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb-group radford-tandem-physics-adamw --wandb-name rei/tandem-wd1e3-ema-adamw --agent rei

# Trial 3: wd=1e-4, no EMA
CUDA_VISIBLE_DEVICES=3 python train.py --dataset tandemfoil --optimizer adamw --lr 5e-4 --cosine-t-max 30 --model-slices 64 --weight-decay 1e-4 --no-use-ema --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --wandb-group radford-tandem-physics-adamw --wandb-name rei/tandem-wd1e4-noema-adamw --agent rei
```

### Peak Memory
- Trial 0 (GPU 0): ~56 GB
- Trial 1 (GPU 1): ~97 GB
- Trial 2 (GPU 2): ~88 GB
- Trial 3 (GPU 3): ~53 GB

### W&B Run IDs
- Trial 0: `eigtav4k`
- Trial 1: `qzfa3cas`
- Trial 2: `4l3kmf9r`
- Trial 3: `at5wwt03`

All runs in group `radford-tandem-physics-adamw`.

### Suggested Follow-ups
1. **Disable EMA by default for TandemFoil** — or at minimum lower ema_decay significantly (e.g. 0.99 or 0.95) and reduce ema_start_step to 0 for short training budgets.
2. **Run Trial 3 recipe with more epochs** — the no-EMA model was still strongly improving at 1 epoch. With a longer budget, val_primary could drop well below 200.
3. **Combine no-EMA with weight decay sweep** — now that EMA is identified as the dominant factor, re-run the wd sweep without EMA for a clean comparison.
4. **Test no-EMA with the Lion optimizer** — the baseline uses Lion; combining no-EMA with Lion could yield further gains.
